### PR TITLE
[Feature] Support NaiveProposer for most cases

### PR DESCRIPTION
**Pull request by huicongyao**
## Motivation

> :bulb: If this PR is a Cherry Pick, the PR title needs to follow the format by adding the [Cherry-Pick] label at the very beginning and appending the original PR ID at the end. For example, [Cherry-Pick][CI] Add check trigger and logic(#5191)

> :bulb: 如若此PR是Cherry Pick，PR标题需遵循格式，在最开始加上[Cherry-Pick]标签，以及最后面加上原PR ID，例如[Cherry-Pick][CI] Add check trigger and logic(#5191)

## Usage or Command

`--speculative-config '{"method": "naive"}'`

## Function Tests

| 测试场景 | 模式 | 测试状态 |
|---------|------|---------|
| 裸FD | 集中式 | ✅ |
| 裸FD | P/D分离 | ✅ |
| 裸FD | 集中式 + logprob | ✅ |
| 裸FD | PD分离 + logprob | ✅ |
| 裸FD | overlap | ✅ |
| RL场景 | naive spec各功能 | 待测试 |

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/model_executor/graph_optimization/cudagraph_piecewise_backend.py [hunk 1]
```diff
@@ -193,7 +193,11 @@ def run_static_model(self, entry: ConcreteSizeEntry, is_decode: bool = False, **
 
     def __call__(self, **kwargs) -> List[paddle.Tensor] | paddle.Tensor:
         # Get real shape (total num tokens)
-        if self.speculative_decoding and all(self.real_bsz_to_captured_size.values()):
+        if (
+            self.speculative_decoding
```
Comment (freeliuzc) [inline fastdeploy/model_executor/graph_optimization/cudagraph_piecewise_backend.py:197]: 这里加是为了兜住什么条件

File: fastdeploy/model_executor/layers/sample/logprobs.py [hunk 2]
```diff
@@ -180,7 +180,16 @@ def build_output_logprobs(
     if is_naive:
         # NAIVE mode: one token per request, logits are already correct
         output_logits = logits
-        token_ids = share_inputs["accept_tokens"][:max_occupied_slots, 0]
+        token_ids = share_inputs["accept_tokens"][:real_bsz, 0]
```
Comment (freeliuzc) [inline fastdeploy/model_executor/layers/sample/logprobs.py]: 这里可以复用投机解码分支吗？

File: fastdeploy/output/token_processor.py [hunk 3]
```diff
@@ -955,8 +955,16 @@ def _process_batch_output(self):
...
                 result.num_input_video_tokens = task.multimodal_inputs.get("num_input_video_tokens", 0)
 
-            if is_prefill and len(token_ids) > 1:
-                result.outputs.draft_token_ids = copy.deepcopy(token_ids)
+            if is_prefill:
+                if len(token_ids) > 1:
+                    result.outputs.draft_token_ids = copy.deepcopy(token_ids)
+                elif (
+                    self.cfg.speculative_config.method == SpecMethod.NAIVE
+                    and self.cfg.scheduler_config.splitwise_role == "prefill"
```
Comment (freeliuzc) [inline fastdeploy/output/token_processor.py]: 这里的判断条件有点乱，建议直接根据方法判断，不能既有 len 又有方法

File: fastdeploy/spec_decode/naive.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+from .base import Proposer
+
+if TYPE_CHECKING:
+    from fastdeploy.config import FDConfig
+
+
+class NaiveProposer(Proposer):
+    """
+    Proposer for NaiveProposer.
+
```
Comment (freeliuzc) [inline fastdeploy/spec_decode/naive.py]: 这里是哪些变量必须使用 proposer 呢，建议删除此结构，不增加冗余代码

File: fastdeploy/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -988,19 +988,28 @@ def insert_tasks_v1(self, req_dicts: BatchRequest, num_running_requests: int = N
...
                         self._cached_real_bsz += 1
                     if self.speculative_decoding:
-                        # D first decode step, [Target first token, MTP first draft token]
-                        # MTP in P only generate one draft token in any num_model_step config
-                        draft_tokens_to_write = request.draft_token_ids[0:2]
-                        if len(draft_tokens_to_write) != 2:
-                            raise ValueError(
-                                "Expected at least 2 draft tokens for speculative suffix decode, "
-                                f"but got {len(draft_tokens_to_write)} for request {request.request_id}."
+                        if self.spec_method in (SpecMethod.MTP, SpecMethod.SUFFIX):
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py]: suffix 暂时不需要传此token，另外即使传，这里也是动态的

File: fastdeploy/model_executor/graph_optimization/cudagraph_piecewise_backend.py (hunk 1 above)
Comment (huicongyao) [reply fastdeploy/model_executor/graph_optimization/cudagraph_piecewise_backend.py:197]: real_bsz_to_captured_size在naive 模式下为空，但是all(self.real_bsz_to_captured_size.values())为true，会导致这块代码运行出错

File: fastdeploy/worker/gpu_model_runner.py (hunk 5 above)
Comment (huicongyao) [reply fastdeploy/worker/gpu_model_runner.py]: 这Naive支持之前，对suffix也有传，去掉的话单测会不通过，此处为了保持一致

File: fastdeploy/spec_decode/naive.py (hunk 4 above)
Comment (huicongyao) [reply fastdeploy/spec_decode/naive.py]: 在initialize_forwardmeta处`self.proposer.fd_config.model_config.moe_phase.phase = "decode"`

File: fastdeploy/output/token_processor.py (hunk 3 above)
Comment (huicongyao) [reply fastdeploy/output/token_processor.py]: 已优化这部分的判断逻辑

File: fastdeploy/model_executor/layers/sample/logprobs.py (hunk 2 above)
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/logprobs.py]: 应该可以

File: fastdeploy/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -2757,12 +2766,14 @@ def _postprocess(
...
-            # will overwrite it with (draft_count + 1) below.
+            # 6. Speculative decode -- proposer run
+            # NAIVE: proposer is None, skip; seq_lens_this_time was
+            # already set to 1 by naive_update_model_status kernel during
+            # sampling. MTP/Ngram: the proposer populates draft_tokens and
+            # updates seq_lens_this_time to (draft_count + 1) for the next
+            # target-model forward pass.
 
-            if self.speculative_decoding and self.proposer is not None:
+            if self.speculative_decoding:
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/worker/gpu_model_runner.py:2799]: 其他硬件暂时不支持naive propose模式

File: fastdeploy/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -988,19 +988,28 @@ def insert_tasks_v1(self, req_dicts: BatchRequest, num_running_requests: int = N
...
-                        # MTP in P only generate one draft token in any num_model_step config
-                        draft_tokens_to_write = request.draft_token_ids[0:2]
-                        if len(draft_tokens_to_write) != 2:
-                            raise ValueError(
-                                "Expected at least 2 draft tokens for speculative suffix decode, "
-                                f"but got {len(draft_tokens_to_write)} for request {request.request_id}."
+                        if self.spec_method in (SpecMethod.MTP, SpecMethod.SUFFIX):
+                            # D first decode step, [Target first token, MTP first draft token]
+                            # MTP in P only generate one draft token in any num_model_step config
+                            draft_tokens_to_write = request.draft_token_ids[0:2]
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py:997]: 这里只有 MTP 是 2，suffix 和 naive 一样是1

File: fastdeploy/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -1493,7 +1502,7 @@ def initialize_forward_meta(self, is_dummy_or_profile_run=False):
         # TODO(wanglongzhi):Modifying the config at runtime is not appropriate; it needs to be moved to forward_meta. It will be used in MoEMethodBase.apply()
         if self.fd_config.parallel_config.use_ep and self.fd_config.scheduler_config.splitwise_role == "mixed":
             self.fd_config.model_config.moe_phase.phase = "decode" if if_only_decode else "prefill"
-            if self.speculative_decoding:
+            if self.speculative_decoding and self.proposer is not None:
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py:1517]: 这里 naive 不走进这个分支吗？

Comment (huicongyao) [reply fastdeploy/worker/gpu_model_runner.py:1517]: 这个分支的代码是为了设置proposer的参数，naive不会使用到

**Review by freeliuzc — APPROVED** [review]

**Review by gongshaotian — APPROVED** [review]
LGTM
